src/Appwrite/Stats/Usage.php

```diff
@@ -0,0 +1,295 @@
+    {
+        $id = \md5("{$time}_{$period}_{$metric}");
+        $this->database->setNamespace('_' . $projectId);
+        try {
+            $document = $this->database->getDocument('stats', $id);
+            if ($document->isEmpty()) {
+                $this->database->createDocument('stats', new Document([
+                    '$id' => $id,
+                    'period' => $period,
+                    'time' => $time,
+                    'metric' => $metric,
+                    'value' => $value,
+                    'type' => $type,
+                ]));
+            } else {
+                $this->database->updateDocument(
+                    'stats',
+                    $document->getId(),
+                    $document->setAttribute('value', $value)
+                );
+            }
+            $this->latestTime[$metric][$period] = $time;
+        } catch (\Exception $e) { // if projects are deleted this might fail
+            if (is_callable($this->errorHandler)) {
+                call_user_func($this->errorHandler, "Unable to save data for project {$projectId} and metric {$metric}: {$e->getMessage()}", $e->getTraceAsString());
+            } else {
+                throw $e;
+            }
+        }
+    }
+
+    /**
+     * Sync From InfluxDB
+     * Sync stats from influxDB to stats collection in the Appwrite database
+     *
+     * @param string $metric
+     * @param array $options
+     * @param array $period
+     *
+     * @return void
+     */
+    private function syncFromInfluxDB(string $metric, array $options, array $period): void
+    {
+        $start = DateTime::createFromFormat('U', \strtotime($period['startTime']))->format(DateTime::RFC3339);
+        if (!empty($this->latestTime[$metric][$period['key']])) {
+            $start = DateTime::createFromFormat('U', $this->latestTime[$metric][$period['key']])->format(DateTime::RFC3339);
+        }
+        $end = DateTime::createFromFormat('U', \strtotime('now'))->format(DateTime::RFC3339);
+
+        $table = $options['table']; //Which influxdb table to query for this metric
+        $groupBy = empty($options['groupBy']) ? '' : ', "' . $options['groupBy'] . '"'; //Some sub level metrics may be grouped by other tags like collectionId, bucketId, etc
+
+        $filters = $options['filters'] ?? []; // Some metrics might have additional filters, like function's status
+        if (!empty($filters)) {
+            $filters = ' AND ' . implode(' AND ', array_map(fn ($filter, $value) => "\"{$filter}\"='{$value}'", array_keys($filters), array_values($filters)));
+        } else {
+            $filters = '';
+        }
+
+        $query = "SELECT sum(value) AS \"value\" FROM \"{$table}\" WHERE \"time\" > '{$start}' AND \"time\" < '{$end}' AND \"metric_type\"='counter' {$filters} GROUP BY time({$period['key']}), \"projectId\" {$groupBy} FILL(null)";
```

Comment:
```suggestion
        $query = "SELECT sum(value) AS \"value\" ";
        $query .= "FROM \"{$table}\" ";
        $query .= "WHERE \"time\" > '{$start}' ";
        $query .= "AND \"time\" < '{$end}' ";
        $query .= "AND \"metric_type\"='counter' {$filters} ";
        $query .= "GROUP BY time({$period['key']}), \"projectId\" {$groupBy} ";
        $query .= "FILL(null)";
```

I know this might be a bit verbose, but I think it might be 100 times easier to read, what do you think?

Reply:
agreed!